File: 2696-the-number-of-beautiful-subsets/the-number-of-beautiful-subsets.cpp
class Solution {
public:
    vector<int> l;
    int dfs(vector<int>& nums,int k,int i){
        if(i==nums.size()) return (l.size()>0)?1:0;
        int m = dfs(nums,k,i+1);
        if(find(l.begin(),l.end(),nums[i]-k) ==l.end()){
            l.push_back(nums[i]);
            m+= dfs(nums,k,i+1);
            l.pop_back();
        }
        return m;
    }
    int beautifulSubsets(vector<int>& nums, int k) {
        sort(nums.begin(),nums.end());
        return dfs(nums,k,0);
    }
};

File: 2696-the-number-of-beautiful-subsets/the-number-of-beautiful-subsets.py
class Solution:
    def beautifulSubsets(self, nums: List[int], k: int) -> int:
        nums.sort()
        d = {}
        l = {}
        s = [-abs(k)]
        def dfs(i):
            if i == len(nums):
                return 1 if len(s)>1 else 0
            q = (tuple(s),i)
            if q in d:return d[q]

            d[q] = dfs(i+1)
            if nums[i] - k not in l:
                s.append(nums[i])
                l[nums[i]] = l.get(nums[i],0) + 1
                d[q] += dfs(i+1)
                if l[nums[i]] == 1:
                    l.pop(nums[i])
                else:
                    l[nums[i]]-=1
                s.pop()
            return d[q]

        a =  dfs(0)
        # print(d)
        return a